### PR TITLE
RFD 101: Kubernetes Per-pod RBAC

### PR DESCRIPTION
**Pull request by tigrato**
[Rendered](https://github.com/gravitational/teleport/blob/tigrato/pod_request_access/rfd/0099-pod-access-requests.md)

### CODE REVIEW

File: rfd/00xx-pod-access-requests.md
Comment (r0mant) [inline rfd/00xx-pod-access-requests.md:98]: Maybe worth clarifying that this is the case for roles that match the cluster your connecting to. So for example of you have:

```
roleA:
  kube_labels: env=prod
  kube_pods: x, y, z

roleB:
  kube_labels: env=stage
```

Then you'll still be able to access anything in "env=stage" clusters and only pods x, y, z in "env=prod" clusters.

Meaning, kube_pods is not a "poison pill" and doesn't apply globally to any roles you have.

File: rfd/00xx-pod-access-requests.md
Comment (r0mant) [inline rfd/00xx-pod-access-requests.md:118]: I would add a brief intro paragraph here that provides comparison with previous approach, something like

```
Second possible approach is, instead of intercepting/rewriting API calls, for Teleport
to manage Kubernetes RBAC policies directly using Kubernetes API and keep relying
on Kubernetes access controls. This will allow us to support access requests but won't
allow to configure role policies in Teleport allowing/denying access to specific pods.
```

And then go into your explanation below.

File: rfd/00xx-pod-access-requests.md
Comment (r0mant) [inline rfd/00xx-pod-access-requests.md:209]: Nit: Technically it doesn't have to. Teleport will let the request through but Kubernetes reject it. I would just say that user's "effective" permissions will be an intersection of what Teleport RBAC and Kubernetes RBAC allow.

File: rfd/00xx-pod-access-requests.md
Comment (r0mant) [inline rfd/00xx-pod-access-requests.md:238]: Same as above: clarify this applies to roles matching the cluster you're accessing. Otherwise it reads like it will have a poison pill behavior.

File: rfd/00xx-pod-access-requests.md
Comment (r0mant) [inline rfd/00xx-pod-access-requests.md:330]: I think we might want to restrict this. Otherwise the user may create pods they don't have access to. Can we parse the request body?

File: rfd/00xx-pod-access-requests.md
Comment (r0mant) [inline rfd/00xx-pod-access-requests.md:347]:
```suggestion
```

**Review by r0mant — COMMENTED** [review]
I've been leaning more and more towards approach 1 with intercepting requests as I've been reading this. Also, as we discussed we could potentially do a combination of 1 and 2 where Teleport can still intercept all requests to pods and dynamically provision Kube groups for access requests. The upside is admin won't have to set them up in Kube separately, the downside is Teleport needs more Kube permissions and we need to implement this synchronization to create/delete them. Maybe mention this more prominently as potential future work.

File: rfd/00xx-pod-access-requests.md
Comment (klizhentas) [inline rfd/00xx-pod-access-requests.md:42]: nit: word `limiting` is used twice

File: rfd/00xx-pod-access-requests.md
Comment (klizhentas) [inline rfd/00xx-pod-access-requests.md:148]: I think this flow will rarely be used, I think we should support the flow most people will use most of the time:

```bash
$ kubectl exec
... teleport intercepts this and creates access request (similarly to how tsh ssh works with on demand access requests)
```

**Review by klizhentas — CHANGES_REQUESTED** [review]
@r0mant @tigrato overall it seems like option 1 (adding teleport-specific access control over existing k8s RBAC and not modifying roles on k8s cluster is the way to go). 

This means that the  K8s role should allow access to pods in the first place, but by default Teleport will restrict exec/port-forwarding commands by default, unless there is a `kubernetes_pods` present that allows this access. This is good for security, because by default exec access to pods will be prohibited unless `kubernetes_pods: ["*"]` for the role. 

Then users would have to request access to this role using our standard flow. This won't be a backwards-compatible behavior change, so it should be role `v6`. The default value will be empty `kubernetes_pods` there.

File: rfd/00xx-pod-access-requests.md
Comment (tigrato) [reply rfd/00xx-pod-access-requests.md:148]: @klizhentas @r0mant 

I am afraid that to implement this flow, we will need to change the current implementation.

Currently, each time a user creates an access request, `tsh` waits for it to be approved and stores the `request_id` in the user's certificate. Only after `tsh` connects to the ssh node.

The problem is that when `kubectl` calls the exec plugin to generate the certificates for the cluster, it does not provide any information about the resource we are trying to access - in this case, the Pod. Thus `tsh kube credentials` cannot request an access request. On the other hand, `kubectl` can't refresh the certificates - it has no idea about it.

I thought of two ways we could solve this problem:

1. The `kube_service` executes the access request as soon as it identifies that the user doesn't have access and:
-    i) sends the command to wait for admin approval. This command refreshes the certificates if the admin approves the request.
-   ii)  Blocks until the request is approved and disconnects. The next time the user calls `tsh kube credentials`, Auth Server will include the `request_id` in the certificate.

3. Teleport stores the `request_ids` server-side - `kube_proxy/service`, and each time the user requests a Kube endpoint, the service inserts the request_ids associated with the active requests for that user. We would need a way to discard the active access requests.

Neither of these is ideal.

File: rfd/00xx-pod-access-requests.md
Comment (tigrato) [reply rfd/00xx-pod-access-requests.md:330]: We can. 

I followed the same criteria used by Kubernetes RBAC which allows you to create pods even if you won't be able to access them.

File: rfd/00xx-pod-access-requests.md
Comment (klizhentas) [reply rfd/00xx-pod-access-requests.md:177]: @tigrato I suggest we go with a much simpler solution, in addition to the flow you created, we can roll out:

`tsh kubectl exec` command, that will mirror and pass through all arguments, except it will detect if the pod will require special access and create ad-hoc access request before pass-through to `kubectl`. Will this work?

Comment (tigrato) [reply rfd/00xx-pod-access-requests.md:177]: @klizhentas 

Yes, this simpler approach will work.

I assume that `tsh kubectl edit pod(s,/, )pod1 pod2 pod3`, and `tsh kubectl port-forward ...` will also have to be supported, right?

I'm a bit wary of using `tsh kubectl` for two reasons:
- tsh would have to catch every command destined for pods - kubectl has a lot of aliases for pods: `po`, `pod`, `pods`, `pod/`.
- For me, it's unnatural to use the tsh prefix before kubectl. We can alias `kubectl="tsh kubectl"` but who would do it?

One possible solution to avoid so many possible combinations is to create a proxy and intercept the `kubectl` HTTP requests destined for pods when `tsh kubectl` is called.

Comment (r0mant) [reply rfd/00xx-pod-access-requests.md:177]: @klizhentas Just wanted to follow up on this ^^^ - how do you feel about the approach with having a local Kubernetes proxy `tsh proxy kube` similar to what we have for database/application access? With this approach it would actually be the proxy that would be requesting access to pods e.g. `tsh proxy kube nginx --namespace=teleport`. A couple nice properties of this approach that I see are:

- We actually have customers asking for a local Kubernetes proxy with ALB support similar to what we did for database access so that would lay a groundwork for that. I think it was actually even considered as this quarter's stretch goal.
- This should integrate nicely with Teleport Connect as it similarly manages local proxies for databases.

So users would start the local proxy (themselves or with Connect) and be able to use `kubectl` to connect through it. We can still keep `tsh kubectl exec` as a single-command convenience wrapper that would start local proxy on demand and make kubectl connect through it (same approach as with `tsh db connect`).

WDYT?

File: rfd/00xx-pod-access-requests.md
Comment (tigrato) [reply rfd/00xx-pod-access-requests.md:83]: Hey @nklaassen,

Good feedback is always welcome!

When a user dynamically requests access to a pod(s), it follows the exact same method you designed for normal resource access requests.
My implementation works as following:
- `tsh request create --resource /<cluster>/pod/<kube_cluster>/<pod_namespace>/<pod_name>`
This request is encoded using `ResourceId` in the following way:
```
types.ResourceId{
   Cluster: "<cluster>",
   Kind: "pod",
   Name: "<kube_cluster>", //does not support wildcards 
   SubResource: "<pod_namespace>/<pod_name>" // both support wildcards matching
}
```

We encode the ResourceId into the user's certificate and we evaluate it at runtime. When Kind is "pod", we need to allow access to the kube_cluster as well and we validate if the pod name is allowed. The resource name is not encoded in the role.

Regarding the `kubernetes_pod` requirement in the role spec, I do imagine the following cases.

1. Admins want to limit access to pods using wildcards or limit access to namespaces without expiration and without creating several Kubernetes RBAC roles.
2. Admins want to limit what users can search using `search_as_role` . I do imagine cases where only a subset of pods are searchable for a certain user/team.

It's not required for dynamic access requests but it's a nice addition tbh.

File: rfd/00xx-pod-access-requests.md
Comment (tigrato) [reply rfd/00xx-pod-access-requests.md:367]: Teleport extracts the `kubernetes_groups/users` from the certificate and sends them as Impersonation headers to the target cluster. The cluster evaluates two things:
- if the auth credentials used have permission to impersonate users and groups. In this case, if teleport credentials allow Impersonation.
- if the Impersonation users and groups' RBAC policies allow the operation.

If teleport allows the resource, but the RBAC denies the request i.e. a user is allowed to access logs for pod X but tries to exec into that Pod, Teleport will let the request through and Kubernetes RBAC will block the request since none of the forwarded impersonation groups and users allow exec.

Once you join a pod, you can directly hit the cluster API but it uses the pod's service account credentials. In the scenario you described, the pod's service account permissions must allow explicitly access to other pods. By default, SA has no permissions. 
If the user tries to impersonate another kube user/group such as `system:masters`, the pod's SA RBAC must explicitly allow impersonation actions. If it doesn't, the request is denied.

The scenario you described depends on what the target pod permissions allow.

File: rfd/00xx-pod-access-requests.md
Comment (tigrato) [reply rfd/00xx-pod-access-requests.md:221]: Good catch. It's a typo.
For roles <=V5, the field is automatically populated with "*" but for a V6 role you must define the value otherwise the request won't return anything since everything will be filtered out.

I need to update the RFD to include the namespace filtering as well - "namespace/*" or "*/somepod-*".

File: rfd/00xx-pod-access-requests.md
Comment (nklaassen) [reply rfd/00xx-pod-access-requests.md:83]: Okay that is making a lot more sense now, thanks for clarifying it for me.

I think it could be made more clear in this doc that `kubernetes_pods` in the role spec is not actually required to implement the pod access requests, if I'm understanding that correctly. And maybe I'm not aware of some motivation for this feature, but my two cents would be to take it out if possible.

Introducing a breaking change to roles and a new major version is adding a lot of scope to this design when I think we could accomplish almost everything here without it. It is also introducing a new concept to our RBAC system, currently if no role has a `deny` for an action, any single role with a matching `allow` is enough, and the access check function actually returns early. With this we will have to check every role for any pod access, so it is a different conceptual model, we can no longer say to users "if any one role allows the action, it is allowed" there will be caveats.

We can always add that feature separately in the future if we want to, but I think you will have a much easier time with pod access requests if you can reduce the scope as much as possible. That's just my thoughts on it.

Comment (tigrato) [reply rfd/00xx-pod-access-requests.md]: @nklaassen 

Sorry for the delay but I missed the notification.

I do agree with you that we do not need the `kubernetes_pods` to implement the Pod Access requests, but from my discussions with @r0mant, this feature must include the Pod RBAC access control (static) + Pod Access requests.
The feature is described in several places with different names and maybe the one I choose for the RFD naming is not the best. I changed the RFD and the PR's name to present a more descriptive name for the feature we want to implement.

Regarding the breaking change: `RoleV6` will be fully compatible with `RoleV5` features and its protobuf will also be compatible so `pre-RoleV6` clients are compatible with the `RoleV6` Auth server.  It requires that the Auth Server downgrades the Role version for clients that do not support version V6 - similarly to the code introduced from `RoleV4` to `RoleV5` code, and in those cases, this feature won't work as expected since the filtering is done in the Kubernetes Service, i.e. costumers must upgrade their services to `>=12.0.0`.

For any role  `<v6`, Teleport will automatically populate `kubernetes_pods` with a wildcard in order to keep the current behavior and to avoid users updating every role in the cluster. `kubernetes_pods` will only be effective for roles with `version: v6`.

Regarding the changes required to Teleport RBAC, I do agree with you that this feature changes the RBAC logic. One possible solution for this is to compute the `kubernetes_groups` and `kubernetes_users` allowed at runtime. Currently, Auth server grabs them when a user requests a cert and embeds them into `tlsca.Identity` based on `kubernetes_labels` matching. 
What we can do is for the Pod endpoint re-compute them based on the request details and forward them to the Kubernetes cluster as Impersonation headers. For the Pod endpoints, Teleport will match the cluster RBAC and Pod RBAC and accumulate the Kubernetes Principals that fulfill the requirements before sending. This removes extra roles that do not match a certain pod and allow users to use different Kubernetes RBAC principals when reaching different pods. An issue will also be present when listing pods since we do not know which Impersonated headers grated access to each resource.

File: rfd/00xx-pod-access-requests.md
Comment (r0mant) [inline rfd/00xx-pod-access-requests.md:202]: Update to `kubernetes_resources`.

File: rfd/00xx-pod-access-requests.md
Comment (r0mant) [inline rfd/00xx-pod-access-requests.md:266]:
```suggestion
$ tsh kubectl exec/logs/edit -it <podName> -n <namespace>
```

File: rfd/00xx-pod-access-requests.md
Comment (r0mant) [inline rfd/00xx-pod-access-requests.md:306]: Update to `kubernetes_resources`.

File: rfd/00xx-pod-access-requests.md
Comment (r0mant) [inline rfd/00xx-pod-access-requests.md:343]: Update to `kubernetes_resources`.

File: rfd/00xx-pod-access-requests.md
Comment (r0mant) [inline rfd/00xx-pod-access-requests.md:130]: Behavior when you have multiple roles granting access to different clusters (via kube_labels) is still not very clear. Do kube_resources matchers apply only to clusters matching kube_labels from the same role? Can you add a section with example when a user has 2 roles.

**Review by r0mant — APPROVED** [review]
lgtm with a few comments

**Review by klizhentas — APPROVED** [review]

File: rfd/00xx-pod-access-requests.md
Comment (ptgott) [reply rfd/00xx-pod-access-requests.md:74]: What would be the value of `kubernetes_resources`? A list of resource URIs like `/apis/GROUP/VERSION/RESOURCETYPE/NAME`?

Comment (tigrato) [reply rfd/00xx-pod-access-requests.md:74]: I forgot to update the RFD's references to `kubernetes_pods` although the RoleV6 PR only has `kubernetes_resources` 

@ptgott  the format is the following:

```yaml
kubernetes_resources:
- kind: pod
   name: "pod1"
   namespace: "namespace1"
- kind: pod
   name: "pod-*"
   namespace: "custom_namespace"

We only support `kind:pod` but it can be extended in the future.

File: rfd/0099-pod-access-requests.md
Comment (r0mant) [inline rfd/0099-pod-access-requests.md:349]: Nit: I'd clarify a bit more.
```suggestion
When accessing a pod in a Kube cluster, Teleport evaluates if any Role matching this Kube cluster allows the access to the Pod,
```

**Review by r0mant — APPROVED** [review]

File: rfd/0101-pod-rbac.md
Comment (SzymonDrosdzol) [inline rfd/0101-pod-rbac.md:598]: Is there anything that can be done to mitigate the risk of leaking env vars and secrets of pods that a given Teleport user doesn't have access to? Such leak feels pretty compromising as it might lead to an entire bypass of the access and visibility limitations that the Teleport is trying to enforce.

At the very least is there a possibility to detect such cases and warn admins when the configuration allows for such bypasses?

File: rfd/0101-pod-rbac.md
Comment (SzymonDrosdzol) [inline rfd/0101-pod-rbac.md:605]: Similarly to the point above: is it possible to at least detect these cases and warn admins that Teleport RBAC can be bypassed in these cases?

File: rfd/0101-pod-rbac.md
Comment (SzymonDrosdzol) [inline rfd/0101-pod-rbac.md:570]: +1 for the "prevent this type of request in Teleport". 

In almost all cases if someone calls DELETE `{namespace_name}/pods` they mean "clear the given namespace", not "delete a few pods that I happen to have access too". So if the user does not have access to all of them, one can assume that the user made some kind of mistake, or doesn't fully understand what he's trying to do. Hence, the safer option is to just reject such operation.

**Review by SzymonDrosdzol — COMMENTED** [review]
Feels like I'm pretty late for the party with my review, as this has been already merged. It's been scheduled this way for Doyensec to review it only now. I wish we had a chance to take part in the discussion earlier.

File: rfd/0101-pod-rbac.md
Comment (tigrato) [reply rfd/0101-pod-rbac.md:598]: I didn't implement this part of the RFD because I have the same concerns.
For now, I implemented a `tsh request search --kind=pod` that does the pod `LIST` request in Teleport AuthServer and returns the `name`, `labels`, and the `namespace` for each pod - not yet merged into master (#19791). We are still discussing if it's the correct approach since it requires Auth Server to connect to a Teleport Proxy. 

With `tsh request search --kind=pod`, we avoid leaking env's and other critical information that, generally, the user must not have access to without requesting access.

File: rfd/0101-pod-rbac.md
Comment (tigrato) [reply rfd/0101-pod-rbac.md:570]: My opinion on this point is not clear.

The main points that make me think it's better to keep this feature are the following:
- from the user's perspective, the namespace only has the pods that he sees.
- delete pods support field selectors and labels which makes sense when filtering the objects to delete.
- I think Kube API has the same behavior, if you only have `delete` permissions for a single pod in the namespace and you try to delete all of them, only one will be deleted 

On the other hand, when we run `kubectl delete pods --all -n namespace`, kubectl does not use that endpoint, instead, it makes a list request and deletes pods 1 by 1.

File: rfd/0101-pod-rbac.md
Comment (SzymonDrosdzol) [reply rfd/0101-pod-rbac.md:598]: @tigrato 
Does implementing the `tsh request search --kind=pod` really prevent the user from running `kubectl get pods -o json|yaml` with the same privs and accessing the dangerous information anyway?

File: rfd/0101-pod-rbac.md
Comment (SzymonDrosdzol) [reply rfd/0101-pod-rbac.md:570]: Got it. Keeping consistence with the Kube API is important too, so I don't have a strong opinion either, after all.

File: rfd/0101-pod-rbac.md
Comment (tigrato) [reply rfd/0101-pod-rbac.md:598]: @SzymonDrosdzol 
Yes.

If you do `kubectl get pods` you just have access to the pods that your role or, in case of an access request, your access request specifies. No other pods will be included.

`tsh request search --kind=pod` will allow you to see what is available in the cluster but will return the pod `name`, `namespace`, and its `labels` hiding every other pod item that might contain secrets or secret refs.

Comment (SzymonDrosdzol) [reply rfd/0101-pod-rbac.md:598]: Got it. Sounds good @tigrato

**Pull request merged**
